### ISSUE TITLE
Support training continuation

### ISSUE THREAD

**Issue by eu9ene** [mozilla/firefox-translations-training#154]
Sometimes things break or we want to fix something in the training config (for example the stopping criteria) and continue model training from the last checkpoint because it's expensive to start from scratch. When we train on slurm+snakemake, the data stays on the disk with all the Marian checkpoints saved. This means we continue training by default when we rerun the pipeline. If we want to start from scratch, we delete the directory. This works even if we change the training corpus and rerun some cleaning scripts.

We should think about how to implement this manual control on TaskCluster with its different caching paradigm. One idea: we can add a config setting like `training: continue` for each of the models. If it says `continue` we preload artifacts for this step from a previous attempt of training for this experiment and language pair (key experiment_name+lang_pair).

**Comment by petemoore** [comment]
Note, under generic worker it is reasonably trivial to mount artifacts that were created in a previous task. See [`task.payload.mounts`](https://github.com/taskcluster/taskcluster/blob/v54.4.2/workers/generic-worker/schemas/multiuser_posix.yml#L274-L283) and [`readOnlyDirectory`](https://github.com/taskcluster/taskcluster/blob/v54.4.2/workers/generic-worker/schemas/multiuser_posix.yml#L680-L743).

Note, the name "read only directory" is a misnomer - the created directory is read/writable. It really should have just been called "directory" or "unretained directory" or something like that. The point is that if you make changes to it, no worker cache will be updated, because after the task completes and artifacts have been uploaded, the local copy is discarded. This is not a problem, because you can still publish the contents of the directory as a directory artifact, or create an archive from the directory (e.g. a tar.gz file or .zip file, or .lz4 file etc) and publish that as an artifact.

**Comment by gregtatum** [comment]
I wonder if the artifacts should be continued based on the hashing of the previous inputs. That way it will automatically happen until we invalidate part of the previous task graph. To me this seems preferable to having a manual `training: continue`. Perhaps we could also have a cache buster property that is incorporated into the hash, so you could manually update that to re-trigger the training.

**Comment by eu9ene** [comment]
We should also think about how to implement fine-tuning of the models that were trained with a different task or even on Snakemake/Slurm. I think we can add an option to download the artifacts from another training run and copy them to a target training folder right before we start training. It can be similar to how it's done for [pre-trained OPUS models ](https://github.com/mozilla/firefox-translations-training/pull/192/files#diff-7487cf799323d7a4d2fe5212612d8fb0b8f00875803950ab7c088db2f86ea14fR20). We can add a URL parameter to config. We should also do the same for the backward model. So we have several use cases that might require a similar solution:
1. Rerun model training from the last checkpoint if the run was terminated
  - If it's an error or the spot instance was terminated, we probably want to continue by default
  - If the job was cancelled we want to provide a user manual control since they might want to start from scratch
2. Download the custom teacher checkpoint from a GCP bucket for fine-tuning (this can mean usage in both `train` and `finetune` steps as they operate on different datasets, but if we use it for `train` the `fine-tune` one obviously should not load anything)
3. Download the custom backward model checkpoint (this can be another student model for example). We do not want to continue training in this case, just start using the model.
4. Download a checkpoint produced by a custom TC task ( in case we want to fine-tune or continue training the model we had in another experiment)

Taking all this into account we might want to just add an extra parameter like a `pretrained-url` and `continue` flag to pair it with for all the training steps and copy artifacts from there before the run starts. Not sure how to make it compatible with TC tasks since the artifacts are not one URL. If we have this we could search for previous tasks and their status on start and modify the config to include URL form another task. 

I would start with the URL approach since it is more generic and would allow us not to lose the work if the training was terminated unexpectedly. Then we can automate it to support the continuation by default. Also, we'll see how often we need this. @gabrielBusta let's sync live on all this before you start implementation

**Comment by bhearsum** [comment]
Is this fixed by #226, or is there more to do? (Or are we keeping it open until we give it some attempts in production?)

**Comment by eu9ene** [comment]
I didn't test it yet but it's supposed to be implemented by @gabrielBusta. Let's close this and create follow up issues if there are bugs.
